eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/AsyncPaymentTriggerer.scala

```diff
@@ -0,0 +1,142 @@
+  private case class WrappedPeerReadyResult(result: PeerReadyNotifier.Result) extends Command
+  private case class WrappedCurrentBlockHeight(currentBlockHeight: CurrentBlockHeight) extends Command
+
+  sealed trait Result
+  case object AsyncPaymentTriggered extends Result
+  case object AsyncPaymentTimeout extends Result
+  case object AsyncPaymentCanceled extends Result
+  // @formatter:on
+
+  def apply(): Behavior[Command] = Behaviors.setup { context =>
+    Behaviors.withMdc(Logs.mdc(category_opt = Some(LogCategory.PAYMENT))) {
+      Behaviors.receiveMessagePartial {
+        case Start(switchboard) => new AsyncPaymentTriggerer(switchboard, context).start()
+      }
+    }
+  }
+}
+
+private class AsyncPaymentTriggerer(switchboard: ActorRef[Switchboard.GetPeerInfo], context: ActorContext[Command]) {
+
+  import AsyncPaymentTriggerer._
+
+  case class Payment(replyTo: ActorRef[Result], timeout: BlockHeight, paymentHash: ByteVector32) {
+    def expired(currentBlockHeight: BlockHeight): Boolean = timeout <= currentBlockHeight
+  }
+
+  case class PeerPayments(notifier: ActorRef[PeerReadyNotifier.Command], pendingPayments: Set[Payment]) {
+    def update(currentBlockHeight: BlockHeight): Option[PeerPayments] = {
+      val expiredPayments = pendingPayments.filter(_.expired(currentBlockHeight))
+      expiredPayments.foreach(e => e.replyTo ! AsyncPaymentTimeout)
+      updatePaymentsOrStop(pendingPayments.removedAll(expiredPayments))
+    }
+
+    def cancel(paymentHash: ByteVector32): Option[PeerPayments] = {
+      val canceledPayment = pendingPayments.find(_.paymentHash == paymentHash)
+      if (canceledPayment.isDefined) canceledPayment.get.replyTo ! AsyncPaymentCanceled
+      updatePaymentsOrStop(pendingPayments.removedAll(canceledPayment))
+    }
+
+    private def updatePaymentsOrStop(pendingPayments: Set[Payment]): Option[PeerPayments] = {
+      if (pendingPayments.isEmpty) {
+        context.stop(notifier)
+        None
+      } else {
+        Some(PeerPayments(notifier, pendingPayments))
+      }
+    }
+
+    def trigger(): Unit = pendingPayments.foreach(e => e.replyTo ! AsyncPaymentTriggered)
+    def timeout(): Unit = pendingPayments.foreach(e => e.replyTo ! AsyncPaymentTimeout)
+  }
+
+  def start(): Behavior[Command] = {
+    context.system.eventStream ! EventStream.Subscribe(context.messageAdapter[CurrentBlockHeight](WrappedCurrentBlockHeight))
+    watching(Map.empty)
+  }
+
+  private def watching(peers: Map[PublicKey, PeerPayments]): Behavior[Command] = {
+    Behaviors.receiveMessagePartial {
+      case Watch(replyTo, remoteNodeId, paymentHash, timeout) =>
+        peers.get(remoteNodeId) match {
+          case None =>
+            val notifier = context.spawn(
+              Behaviors.supervise(PeerReadyNotifier(remoteNodeId, switchboard, None)).onFailure(SupervisorStrategy.stop),
+              s"peer-ready-notifier-$remoteNodeId",
+            )
```

Comment:
I think it would be safer to spawn anonymous actors here:
```suggestion
            val notifier = context.spawnAnonymous(Behaviors.supervise(PeerReadyNotifier(remoteNodeId, switchboard, None)).onFailure(SupervisorStrategy.stop))
```

Otherwise we can run into the following race condition, which would be very rare, but would crash the actor because of a child actor name conflict:

- we successfully wake up a peer: in the handler for `PeerReadyNotifier.PeerReady` we remove it from our `peers` map, but we do so before the child actor has stopped itself
- another request comes in for the same peer: it's not in our map, but the child actor is still alive so we fail to create a new one because the name is already taken

Reply:
Good point, no reason to give it a specific name; fixed it in [9619b6e](https://github.com/ACINQ/eclair/pull/2491/commits/9619b6e3ee7c59cdfd93d3630dcce7a55d54c567)